runtime/lua/_vim9script.lua

```diff
@@ -0,0 +1,627 @@
+-------------------------------------------------------------------------------
+-- This file is auto generated by vim9jit. Do not edit by hand.
+--  All content is in the source repository.
+--  Bugs should be reported to: github.com/tjdevries/vim9jit
+--
+-- In addition, this file is considered "private" by neovim. You should
+-- not expect any of the APIs, functions, etc to be stable. They are subject
+-- to change at any time.
+-------------------------------------------------------------------------------
+
+local vim9 = (function()
+  local M = {}
+
+  M.ternary = function(cond, if_true, if_false)
+    if cond then
+      if type(if_true) == 'function' then
+        return if_true()
+      else
+        return if_true
+      end
+    else
+      if type(if_false) == 'function' then
+        return if_false()
+      else
+        return if_false
+      end
+    end
+  end
+
+  M.fn_mut = function(name, args, info)
+    local result = vim.fn._Vim9ScriptFn(name, args)
+    for idx, val in pairs(result[2]) do
+      M.replace(args[idx], val)
+    end
+
+    -- Substitute returning the reference to the
+    -- returned value
+    if info.replace then
+      return args[info.replace + 1]
+    end
+
+    return result[1]
+  end
+
+  M.replace = function(orig, new)
+    if type(orig) == 'table' and type(new) == 'table' then
+      for k in pairs(orig) do
+        orig[k] = nil
+      end
+
+      for k, v in pairs(new) do
+        orig[k] = v
+      end
+
+      return orig
+    end
+
+    return new
+  end
+
+  M.index = function(obj, idx)
+    if vim.tbl_islist(obj) then
+      if idx < 0 then
+        return obj[#obj + idx + 1]
+      else
+        return obj[idx + 1]
+      end
+    elseif type(obj) == 'table' then
+      return obj[idx]
+    elseif type(obj) == 'string' then
+      return string.sub(obj, idx + 1, idx + 1)
+    end
+
+    error('invalid type for indexing: ' .. vim.inspect(obj))
+  end
+
+  M.index_expr = function(idx)
+    if type(idx) == 'string' then
+      return idx
+    elseif type(idx) == 'number' then
+      return idx + 1
+    else
+      error(string.format('not yet handled: %s', vim.inspect(idx)))
+    end
+  end
+
+  M.slice = function(obj, start, finish)
```

Comment:
Would it be much work to add `@param` annotations with the type? Or would it be too difficult since vim9 types don't directly map to Lua types?

Reply:
what would be the purpose? they are all private functions

Reply:
(and they basically only get called from generated code [which is just lua code generated from the rust stuffz])

Reply:
I guess there wouldn't be much purpose. Just to make the generated code a bit more complete and readable if anything.

Reply:
it could definitely be useful if we were to use vim9jit as a starting point for modules that we later "own". Assuming that the impl cost isn't too high.

Reply:
I can do it in future versions, don't know that we need to wait for those in this PR

Reply:
Just an open suggestion. Certainly not a requirement for this PR.